src/README.md

```diff
@@ -303,15 +306,11 @@ Currently, every `Environment` class is associated with:
 
 * One [event loop][]
 * One [`Isolate`][]
-* One main [`Context`][]
+* One principal [`Realm`][]
 
 The `Environment` class contains a large number of different fields for
-different Node.js modules, for example a libuv timer for `setTimeout()` or
-the memory for a `Float64Array` that the `fs` module uses for storing data
-returned from a `fs.stat()` call.
-
-It also provides [cleanup hooks][] and maintains a list of [`BaseObject`][]
-instances.
+different built-in modules that can be shared across different `Realm`
+instances, for example a libuv timer for `setTimeout()`.
```

Comment:
Thanks for pointing this out! I've added a list of examples that can be shared across realms with an `Environment`. 

It is still yet to be done to share the async hook info between the realms -- it is necessary to link the async continuation between realm boundaries for AsyncLocalStorage to propagate correctly. This is essential to allow JS object access between multiple execution "environments" on the same thread.

Reply:
I’m not really sure that this answers the question. How is the end state of `Realm` going to be different from what `Environment` is (or used to be before `Realm`s were introduced)?

Reply:
The difference between the end state of the `Realm` and the `Environment` would be:
1. An `Environment` is associated with an `Isolate` and provides the per-isolate hooks and APIs, and per-thread states, for instance, inspector agents, profilers, and APIs like `RequestInterrupt()` and `can_call_into_js()`. 
2. A `Realm` is associated with a `Context` and consists of a global object and can be extended as a principal realm or a synthetic realm.

As realms share the `Environment` instance, we don't have to create inspector IO threads, profiler connections for each realm. The async local storage needs be propagated across async boundaries in another `Realm` of an `Environment` too.

Additionally, a `Realm` must be able to be repetitively created on a same `Isolate` and weakly referenced to properly support the ECMAScript `ShadowRealm` API.

As a conclusion, it is necessary to split the `Realm` and `Environment` do distinguish per-isolate states and per-context states.

Reply:
> 1. An `Environment` is associated with an `Isolate` and provides the per-isolate hooks and APIs,

You’re describing `IsolateData` here, though.

Reply:
> You’re describing `IsolateData` here, though.

Yes, it's true that `IsolateData` is also associated with an `Isolate`. However, as documented, `IsolateData` acts like a string table and contains information (e.g. startup snapshot data) about the isolate. It has its distinct properties compared to an `Environment`.

Compared to `IsolateData`, `Environment` at the end state contains the necessary handles (bootstrapped with the principal realm) to propagate events across realm boundaries, for instance, propagating async local storages and promise rejection events from `ShadowRealm`.

Reply:
I think the current state we have is:

1. There is IsolateData which is meant to hold per-isolate data
2. There are Realms with are meant to hold per-context data
3. There is Environment which has been a hotch-potch structure that people dump everything on since forever, so it contains both per-isolate data and per-context data in practice.

What we've been trying to do is to move per-isolate data in Environment to IsolateData and per-context data to Realm, it's happening but it's really not there yet.

Reply:
@addaleax do you think your question has been answered?